macros/src/derive_fn.rs

```diff
@@ -136,15 +145,50 @@ pub fn derive_fn(
         #link_section
         pub static #spec_ident: [u8; #spec_xdr_len] = *#spec_xdr_lit;
 
-        #export_name
-        pub fn #wrap_ident(env: stellar_contract_sdk::Env, #(#wrap_args),*) -> stellar_contract_sdk::RawVal {
-            <_ as stellar_contract_sdk::IntoVal<stellar_contract_sdk::Env, stellar_contract_sdk::RawVal>>::into_val(
-                #call(
-                    #env_call
-                    #(#wrap_calls),*
-                ),
-                &env
-            )
+        pub mod #mod_ident {
+            #export_name
+            pub fn call(env: stellar_contract_sdk::Env, #(#wrap_args),*) -> stellar_contract_sdk::RawVal {
+                #use_trait;
+                <_ as stellar_contract_sdk::IntoVal<stellar_contract_sdk::Env, stellar_contract_sdk::RawVal>>::into_val(
+                    #call(
+                        #env_call
+                        #(#wrap_calls),*
+                    ),
+                    &env
+                )
+            }
+
+            pub fn call_slice(env: stellar_contract_sdk::Env, args: &[stellar_contract_sdk::RawVal]) -> stellar_contract_sdk::RawVal {
+                call(env, #(#slice_args),*)
+            }
+        }
+    }
+}
+
+#[allow(clippy::too_many_lines)]
+pub fn derive_add_functions<'a>(
+    ty: &Box<Type>,
+    methods: impl Iterator<Item = &'a syn::ImplItemMethod>,
+    feature: &Option<String>,
+) -> TokenStream2 {
+    let (idents, wrap_idents): (Vec<_>, Vec<_>) = methods
+        .map(|m| {
+            let ident = format!("{}", m.sig.ident);
+            let wrap_ident = format_ident!("__{}", m.sig.ident);
+            (ident, wrap_ident)
+        })
+        .multiunzip();
+    let cfg = if let Some(cfg_feature) = feature {
+        quote! { #[cfg(feature = #cfg_feature)] }
+    } else {
+        quote! {}
+    };
+    quote! {
+        #cfg
+        impl stellar_contract_sdk::AddFunctions for #ty {
+            fn add_functions(tc: &mut stellar_contract_sdk::TestContract) {
```

Comment:
Could we change this `AddFunctions` trait to `FunctionSet` and have it simply return a list of functions?

The reason I'm suggesting this is that right now this trait and function are bound tightly to `TestContract`, and have this almost-cyclic relationship with it because we give it a TC, and then it calls the TC. The developer calls `my_contract.add_functions(tc)`. I think there are some advantages to presenting it as `tc.add_functions(my_contract.function_set())`, where here we simply provide a function that gives the developer access to the functions that exist, and that can then be hooked up with the TC.

If there are any downsides to this feel free to disregards.

Reply:
I took a totally different approach. `TestContract` is gone. SDK exports a modified version of `ContractFunctionSet` that takes `Env` instead of `&EnvImpl` (to avoid exposing `EnvImpl` to clients). The macro generates an implementation of `ContractFunctionSet`, and then registration is very easy and smooth.

This can easily be extended to support multiple function sets by extending `InternalContractFunctionSet` to contain multiple `ContractFunctionSet`s and chaining them, because they return `Option<...>`.